lib/pengine/group.c

```diff
@@ -187,32 +187,38 @@ pe__group_xml(pcmk__output_t *out, va_list args)
     GListPtr only_rsc = va_arg(args, GListPtr);
 
     GListPtr gIter = rsc->children;
-    char *count = crm_itoa(g_list_length(gIter));
 
     int rc = pcmk_rc_no_output;
-    gboolean printed_header = FALSE;
+    gboolean print_everything = TRUE;
 
     if (rsc->fns->is_filtered(rsc, only_rsc, TRUE)) {
         return rc;
     }
 
+    print_everything = pcmk__str_in_list(only_rsc, rsc_printable_id(rsc)) ||
+                       (strstr(rsc->id, ":") != NULL && pcmk__str_in_list(only_rsc, rsc->id));
+
     for (; gIter != NULL; gIter = gIter->next) {
         pe_resource_t *child_rsc = (pe_resource_t *) gIter->data;
 
-        if (!printed_header) {
-            printed_header = TRUE;
+        if (child_rsc->fns->is_filtered(child_rsc, only_rsc, print_everything)) {
+            continue;
+        }
 
+        if (rc == pcmk_rc_no_output) {
+            char *count = crm_itoa(g_list_length(gIter));
```

Comment:
Nope, just some sloppy code rearranging.  Fixed.

Reply:
Although, putting that back where it was results in this weird situation:
```
  <resources>
    <group id="exim-group" number_resources="2">
      <resource id="Email" resource_agent="lsb:exim" role="Started" active="true" orphaned="false" blocked="false" managed="true" failed="false" failure_ignored="false" nodes_running_on="1">
        <node name="cluster02" id="2" cached="true"/>
      </resource>
    </group>
  </resources>
```

That's because the number of resources is calculated before we do any filtering, so we don't actually know how many resources are going to be output.  We'd probably prefer that number reflect the actual output, not the number of resources in the group overall, right?  In that case, I'd need to walk the group twice - once to count, and once to output.  That means calling is_filtered twice for everything.  I suppose for very large groups,that could be a performance hit (though only when we're outputting things, not in critical code).

Reply:
Actually I think we do want number_resources to be the number in the whole group. It's an attribute of the group itself, not of the output

Reply:
Okay, that's certainly a lot easier to do.